connectors/src/connectors/microsoft/temporal/activities.ts

```diff
@@ -648,6 +675,137 @@ export async function syncOneFile({
+    throw new Error(`Configuration for connector ${connectorId} not found`);
+  }
+
+  const dataSourceConfig = dataSourceConfigFromConnector(connector);
+
+  const { nodeType } = typeAndPathFromInternalId(nodeId);
+
+  if (nodeType !== "drive" && nodeType !== "folder") {
+    throw new Error(`Node ${nodeId} is not a drive or folder`);
+  }
+
+  const node = await MicrosoftNodeResource.fetchByInternalId(
+    connectorId,
+    nodeId
+  );
+
+  if (!node) {
+    throw new Error(`Root node resource ${nodeId} not found`);
+  }
+
+  const client = await getClient(connector.connectionId);
+
+  if (!node.deltaLink) {
+    throw new Error(
+      `Delta link not found for root node resource ${JSON.stringify(node.toJSON())}`
+    );
+  }
+
+  logger.info({ connectorId, node }, "Syncing delta for node");
+
+  // Goes through pagination to return all delta results. This is because delta
+  // list can include same item more than once and api recommendation is to
+  // ignore all but the last one.
+  //
+
+  // although the list might be long, this should not be an issue since in case
+  // of activity retry, files already synced won't be synced again thanks to the
+  // lastSeenTs check VS startSyncTs
+  //
+  // If it ever becomes an issue, redis-caching the list and having activities
+  // grabbing pages of it can be implemented
+  const { results, deltaLink } = await getFullDeltaResults(
+    client,
+    nodeId,
+    node.deltaLink
+  );
+  const uniqueDriveItemList = removeDuplicateItems(results);
+
+  for (const driveItem of uniqueDriveItemList) {
+    heartbeat();
+
+    if (!driveItem.parentReference) {
+      throw new Error(`Unexpected: parent reference missing: ${driveItem}`);
+    }
+
+    const internalId = getDriveItemInternalId(driveItem);
+
+    if (driveItem.file) {
+      if (driveItem.deleted) {
+        await deleteFile({ connectorId, internalId, dataSourceConfig });
+      } else {
+        await syncOneFile({
+          connectorId,
+          dataSourceConfig,
+          providerConfig,
+          file: driveItem,
+          parentInternalId: getParentReferenceInternalId(
+            driveItem.parentReference
+          ),
+          startSyncTs,
+        });
+      }
+    } else if (driveItem.folder) {
+      if (driveItem.deleted) {
+        // no need to delete children here since they will all be listed
+        // in the delta with the 'deleted' field set
+        await deleteFolder({ connectorId, internalId });
+      } else {
+        await MicrosoftNodeResource.updateOrCreate(
+          connectorId,
+          itemToMicrosoftNode("folder", driveItem)
+        );
+      }
+    } else {
+      throw new Error(`Unexpected: driveItem is neither file nor folder`);
+    }
+  }
+
+  await node.updateDeltaLink(deltaLink);
+}
+
+/**
+ *  As per recommendation, remove all but the last occurences of the same
+ *  driveItem in the list
+ */
+function removeDuplicateItems(deltaList: microsoftgraph.DriveItem[]) {
```

Comment:
Name is a bit confusing, what about something like `getLatestOccurences`?

Reply:
Noob question: I guess `/detla` returns `DriveItem` sorted by creation date?

Reply:
microsoft api just says "use the last occurence" but also says "the order is not defined and could be anything" :)